icu4c/source/common/unicode/uset.h

```diff
@@ -1285,4 +1286,95 @@ U_CAPI UBool U_EXPORT2
 uset_getSerializedRange(const USerializedSet* set, int32_t rangeIndex,
                         UChar32* pStart, UChar32* pEnd);
 
-#endif
+#if U_SHOW_CPLUSPLUS_API
+#ifndef U_HIDE_DRAFT_API
+
+namespace U_HEADER_ONLY_NAMESPACE {
+
+// Note: Not U_COMMON_API, and not a subclass of UMemory, because this is a header-only class,
+// not intended to be used via export from the ICU DLL.
+
+// Note: I think I am trying to build something like an
+// https://en.cppreference.com/w/cpp/named_req/InputIterator
+
+/**
+ * TODO API docs
+ * @draft ICU 76
+ */
+class UnicodeSetRangeIterator {
+public:
+    UnicodeSetRangeIterator(const UnicodeSetRangeIterator &other) = default;
+
+    bool operator==(const UnicodeSetRangeIterator &other) const {
+        // same set, and both or neither are done, and same index if not done
+        return uset == other.uset &&
+            (rangeIndex >= rangeCount) == (other.rangeIndex >= other.rangeCount) &&
+            ((rangeIndex >= rangeCount) || rangeIndex == other.rangeIndex);
+    }
+
+    bool operator!=(const UnicodeSetRangeIterator &other) const { return !operator==(other); }
+
+    std::pair<UChar32, UChar32> operator*() const {
```

Comment:
My immediate thought is that `std::pair<UChar32, UChar32>` is too nondescript a type for a range of code points.
I would expect it to be, well, a [range](https://en.cppreference.com/w/cpp/ranges/range).
In particular, I would expect to be able to iterate over it:
```C++
for (auto const range: set.ranges()) {
  if (MayContainCats(range)) {
    for (UChar32 const cp : range) {
      if (u_iscat(cp)) {
        u_pet(cp);
      }
    }
  }
}
```

There is a bit of boilerplate needed to get that, but I just remembered that conveniently some egg wrote something like it back in October: https://github.com/unicode-org/unicodetools/blob/ce2ff440fca747e32e028594d8484db6e1d33afb/c/uca/sifter/ucd.hpp#L29-L91.

Reply:
> My immediate thought is that `std::pair<UChar32, UChar32>` is too nondescript a type for a range of code points. I would expect it to be, well, a [range](https://en.cppreference.com/w/cpp/ranges/range).

Makes sense, and it's tempting, but then either I can't use the nifty `for (auto [start, end] : set.ranges())` loop, or else I have to make funny range field names public. (See below.)

I also don't see why someone would want to do the double loop -- the next thing I am going to add is a code point iterator. Just like in Java, we give easy iteration over either code points or ranges, and callers should choose what's most useful to them. They should not need a double loop.

```
class CodePointRange {
public:
    CodePointRange(UChar32 start, UChar32 end) : start_(start), end_(end) {}
    CodePointRange(const CodePointRange &other) = default;
    inline operator std::pair<UChar32, UChar32>() const { return std::make_pair(start_, end_); }
    UChar32 begin() const { return start_; }
    UChar32 end() const { return end_ + 1; }

private:
    UChar32 start_, end_;
};

usettest.cpp:4274:16: error: cannot decompose private member 'start_' of 'icu_76::header::CodePointRange'
    for (auto [start, end] : set.ranges()) {
```

Reply:
> but then either I can't use the nifty for (auto [start, end] : set.ranges()) loop, or else I have to make funny range field names public. (See below.)

I think _especially_ If you want to allow this nifty syntax, you should use a proper type, and then those member variables can have names and comments to clarify their semantics (iirc ICU and standard C++ have different opinions on what end means?). I see no problem with making it a struct with two public member variables while also giving it the structure of a range.

As for the use case, that double loop was a toy example, but being able to treat it as a range, and to feed it to various things that operate on ranges, seems useful. (certainly having a size() seems convenient).

Reply:
> iirc ICU and standard C++ have different opinions on what end means

Indeed looking at your example above
```C++
    for (auto [start, end] : set.ranges()) {
        printf("set.range U+%04lx..U+%04lx\n", (long)start, (long)end);
    }
```
it looks like those are inclusive ranges, which are not the norm in C++. This is fine, since those _are_ the norm in Unicode, but then they really want to be their own type, not a nondescript pair.
The member variables can probably be called first and last or somesuch to avoid the word `end` which would be interpreted as something else here.

Reply:
Ok, done. Added CodePointRange and got it to work.
Note that in ICU, especially in UnicodeSet/USet, ranges have _always_ been inclusive start..end.